tests/feature/Issues/Issue1427Test.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/*
+ * Copyright (c) Alexandre Gomes Gaigalas <alganet@gmail.com>
+ * SPDX-License-Identifier: MIT
+ */
+
+declare(strict_types=1);
+
+test('https://github.com/Respect/Validation/discussions/1427', expectAll(
+    fn () => v::each(
+        v::arrayVal()
+            ->key('groups', v::each(v::intVal()))
+            ->key('permissions', v::each(v::boolVal()))
+    )
+        ->assert([
+            16 => [
+                'groups' => [1, 'A', 3, 4, 5],
+                'permissions' => [
+                    'perm1' => true,
+                    'perm2' => false,
+                    'perm3' => 'boom!',
+                ],
+            ],
+            18 => false,
+            24 => ['permissions' => false],
+        ]),
+    '`.16.groups.1` must be an integer value',
+    <<<'FULL_MESSAGE'
+    - Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid
+      - `.16` must pass the rules
+        - Each item in `.groups` must be valid
+          - `.1` must be an integer value
+        - Each item in `.permissions` must be valid
+          - `.perm3` must be a boolean value
+      - `.18` must pass all the rules
+        - `.18` must be an array value
+        - `.groups` must be present
+        - `.permissions` must be present
+      - `.24` must pass the rules
+        - `.groups` must be present
+        - `.permissions` must be iterable
+    FULL_MESSAGE,
+    [
+        'messages' => ['each' => 'Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid'],
+        'children' => [
+            16 => [
+                'messages' => ['allOf' => '`["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]` must pass the rules'],
+                'children' => [
+                    'groups' => [
+                        'messages' => ['each' => 'Each item in `[1, "A", 3, 4, 5]` must be valid'],
+                        'children' => [
+                            1 => [
+                                'messages' => ['intVal' => '"A" must be an integer value'],
+                            ],
+                        ],
+                    ],
+                    'permissions' => [
+                        'messages' => ['each' => 'Each item in `["perm1": true, "perm2": false, "perm3": "boom!"]` must be valid'],
+                        'children' => [
+                            'perm3' => [
+                                'messages' => ['boolVal' => '"boom!" must be a boolean value'],
```

Comment:
I'm not replacing the input of each rule with the key, because I think that, since the structure is already in the array, it's best to keep the original message instead, so you can actually see the value that failed.